torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -138,53 +132,56 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
     future->markCompleted();
   }
 
-  enqueue(SendWork(dstRank, std::move(message)));
+  enqueueSend(RpcWork(dstRank, std::move(message)));
   return future;
 }
 
-void ProcessGroupAgent::enqueue(SendWork work) {
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-  sendQueue_.emplace_back(std::move(work));
-  lock.unlock();
-
-  workProduceCV_.notify_one();
+void ProcessGroupAgent::enqueueSend(RpcWork work) {
+  threadPool_.run(std::bind(
+    [&](const RpcWork& work) {
+      std::stringstream ss;
+      serialize(work.message_, ss);
+      std::string str = ss.str();
+
+      std::vector<torch::Tensor> preamble = {
+        torch::tensor(
+          {
+            (int64_t)pg_->getRank(),
+            (int64_t)str.length(),
+          }, {torch::kLong})
+      };
+
+      // ProcessGroup is not thread-safe when sending with the same tag, hence
+      // the lock
+      std::unique_lock<std::mutex> lock(sendMutexes_.get()[work.rank_]);
+      pg_->send(preamble, work.rank_, work.rank_ /* channelTag */)->wait();
```

Comment:
Do we need to wait under the lock? This would be pretty expensive since the default timeout for ProcessGroupGloo is 10s.

Reply:
I assume so, otherwise, the `preamble` and the `payload` tensors could be destructed before the send finish. And that would load to nondeterministic behavior I think.

Reply:
ProcessGroupGloo returns `SendWork` which captures the tensor to ensure it doesn't get deallocated. We should be able to write something like this without any issues:

```
std::unique_loc<> lock(...)
auto work = pg_->send(....)
lock.unlock()
work.wait()
```